util/recorder.js

```diff
@@ -0,0 +1,952 @@
+      reqresp.expectedSize = Number(resp.headers.get("content-length") || -1);
+    }
+
+    if (reqresp.expectedSize === 0) {
+      reqresp.payload = new Uint8Array();
+      return;
+
+    } else if (!resp.body) {
+      logger.error("Empty body, stopping fetch", {url}, "recorder");
+      await this.recorder.crawlState.removeDupe(ASYNC_FETCH_DUPE_KEY, url);
+      return;
+    }
+
+    reqresp.fillFetchResponse(resp);
+
+    return this.takeReader(resp.body.getReader());
+  }
+
+  async* takeReader(reader) {
+    try {
+      while (true) {
+        const { value, done } = await reader.read();
+        if (done) {
+          break;
+        }
+
+        yield value;
+      }
+    } catch (e) {
+      logger.warn("takeReader interrupted", {...errJSON(e), url: this.reqresp.url, ...this.recorder.logDetails}, "recorder");
+      this.reqresp.truncated = "disconnect";
+    }
+  }
+
+  async* takeStreamIter(cdp, stream) {
+    try {
+      while (true) {
+        const {data, base64Encoded, eof} = await cdp.send("IO.read", {handle: stream});
+        const buff = Buffer.from(data, base64Encoded ? "base64" : "utf-8");
+
+        yield buff;
+
+        if (eof) {
+          break;
+        }
+      }
+    } catch (e) {
+      logger.warn("takeStream interrupted", {...errJSON(e), url: this.reqresp.url, ...this.recorder.logDetails}, "recorder");
+      this.reqresp.truncated = "disconnect";
+    }
+  }
+}
+
+// =================================================================
+class ResponseStreamAsyncFetcher extends AsyncFetcher
+{
+  constructor(opts) {
+    super(opts);
+    this.cdp = opts.cdp;
+    this.requestId = opts.requestId;
+  }
+
+  async _doFetch() {
+    const { requestId, reqresp, cdp } = this;
+    const { url } = reqresp;
+    logger.debug("Async started: takeStream", {url}, "recorder");
+
+    const { stream } = await cdp.send("Fetch.takeResponseBodyAsStream", {requestId});
+
+    return this.takeStreamIter(cdp, stream);
+  }
+}
+
+// =================================================================
+class NetworkLoadStreamAsyncFetcher extends AsyncFetcher
+{
+  constructor(opts) {
+    super(opts);
+    this.cdp = opts.cdp;
+  }
+
+  async _doFetch() {
+    const { reqresp, cdp } = this;
+    const { url } = reqresp;
+    logger.debug("Async started: loadNetworkResource", {url}, "recorder");
+
+    const options = {disableCache: false, includeCredentials: true};
+
+    let result = null;
+
+    try {
+      result = await cdp.send("Network.loadNetworkResource", {frameId: reqresp.frameId, url, options});
+    } catch (e) {
+      logger.debug("Network.loadNetworkResource failed, attempting node fetch", {url, ...errJSON(e), ...this.recorder.logDetails}, "recorder");
+      return await super._doFetch();
+    }
+
+    const { stream, headers, httpStatusCode, success, netError, netErrorName } = result.resource;
+
+    if (!success || !stream) {
+      //await this.recorder.crawlState.removeDupe(ASYNC_FETCH_DUPE_KEY, url);
```

Comment:
```suggestion
```
Seems out of place?

Reply:
Maybe keep for now as still want to revisit how dupes are handled. The idea was that if it failed, we wouldn't track it as a dupe..